### PR TITLE
Support loading configuration from a config file.

### PR DESCRIPTION
**Pull request by Urik**
Added an option to load ignore rules from a lines separated file by using the "--ignores-file"
CLI flag.

Usage would be `depcheck --ignores-file="pathToFile"`

I wanted to add unit tests but could find no mocking libraries installed in the system to handle "rs.readFile" calls, nor was I sure of exactly where should I add my unit tests.
Any help in that front will be very welcome.

### CODE REVIEW

File: src/utils/read-lines-separated-file.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { readFile } from 'fs';
+
+export default function readLinesSeaparatedFile(filePath) {
+  return new Promise((resolve, reject) => {
+    readFile(filePath, 'utf-8', (err, content) => {
+      if (err) {
+        return reject(err);
+      }
+
+      return resolve(content.split('\n'));
```
Comment (rumpl) [inline src/utils/read-lines-separated-file.js]: Could we use a regular expression here, on windows a new line is `\r\n`.

**Comment by rumpl** [comment]
Hey @Urik 

Thanks for the PR, I'll try to find some time today and see how we could add unit tests for this.

**Comment by rumpl** [comment]
@Urik I just pushed a test for this [feature in another branch](https://github.com/depcheck/depcheck/commit/1efb6dead0e07ed7dcae37830c923202b4dc0aab), you can copy/paste the test in your branch and then we'll be ready to merge :)

File: src/utils/read-lines-separated-file.js (hunk 1 above)
Comment (Urik) [reply src/utils/read-lines-separated-file.js]: I was under the impression fs.readLine always interprets `\r\n` as simply `\n` but now I couldn't find anything about it. Fixed.

**Comment by Urik** [comment]
Thank you for the tests!
Just added them, so the code should be ready for review again.

**Comment by rumpl** [comment]
This looks good and I'm really close to merging it but it got me thinking, why stop here?

There were people asking for a .depcheckrc (or whatever) file so that it's more CI friendly.

What do you think @Urik ? Are you motivated enough to change this PR to a full-fledged depcheck config file? I don't think it would be too much work...

**Comment by Urik** [comment]
Mmmh I agree it'd be a very nice addition and a .depcheckrc file would make life much easier for me for sure, but it would still be nice to provide the option to have the ignore rules in a separate file, away from other config options, making it easier to separate the concern of the ignore rules from the rest of the configuration.

I can work on it on the following weekends, but it'll take me some time as I'm very busy on the weekdays and it'll take a few hours. I think it'd be best handled as a separate feature and pull request.

**Comment by rumpl** [comment]
@KaboomFox @lijunle @mnkhouri @43081j 

Any thoughts on this?

**Comment by 43081j** [comment]
in most projects ive seen, there wouldn't usually be a "load ignores from file" feature.

normally you'd have a `*rc.{json,yml}` file which can contain a `ignore: [a, b, c]` while the CLI equivalent is `--ignore=a,b,c`.

personally id prefer this approach rather than supporting loading the list from its own standalone file, for sake of keeping it well organised and to a 'convention' (in that everything is guaranteed to be on the command line or in a config file, nowhere else).

but if you're all up for having this feature, i'd do it along side the rc file implementation (json only in first cut).

**Comment by mnkhouri** [comment]
I'm also in agreement: if we have some configuration available via a file (as this PR implements, which I think is a great idea!) we should have all configuration eventually available via file.

I've seen a lot of other projects use [cosmiconfig](https://github.com/davidtheclark/cosmiconfig), which implements the logic for finding and parsing config files. It seems pretty standard among other javascript cli tools (e.g. babel, eslint, prettier).

I've never personally developed with cosmiconfig, but the documentation is good and the package is well-tested, so perhaps it would be a reasonable lift to make use of cosmiconfig with this PR? We could start with just supporting `ignores` in the config file, then expand in other PRs

**Comment by Urik** [comment]
My rationale was that ignores rules might change more often than not while config values wouldn't, thus it would be nice to be able to easily tell away changes between a proper config value and one of the possibly many ignores. 

I like @mnkhouri 's approach of adding support for other settings progressively.

**Comment by lijunle** [comment]
I like the `*rc` approach in general. It will be very help if @Urik can make that work.

Besides, may I step back and ask what kind of packages do you want to ignore? You said that it is related to unit test, but with what concrete case?

**Comment by Urik** [comment]
I finally had a free Sunday and implemented the RC file functionality. Thanks @mnkhouri for the cosmiconfig suggestion, it was of great help.

Something I haven't had time to implement yet is supporting comma separated lists in CLI arguments, and real JSON arrays in the .depcheckrc file, since that'd mean changing more parts of the logic and I had no time.

Is it something you want supported, or would you like to keep the comma separated notation for the sake of consistency and simplicity?

**Comment by Urik** [comment]
Sorry it took me so long, I finally had time today to implement this. 
The feature is ready for review. It'll mimick the cli args, meaning you can use the same arguments as in the CLI arguments. 
#### Priority
If an argument is found both in the CLI and in the .depcheckrc file, the CLI argument will take precedence, overwriting the value in the .depcheckrc.
#### Format
the .depcheckrc file can use YAML notation:    
```
ignores: ["a", "b"]
ignore-dirs:
  - folderA
  - folderB
```
or also JSON notation:
```
{
  "ignores": ["a", "b"],
  "ignore-dirs": ["folderA", "folderB"]
}
```

Also to take into account, keys can be `camelCase`, or also `kebab-case`, it'll have no effect.

#### Extension
Thanks to cosmiconfig, the .depcheckrc file can have no extension, `.json` extension, `.yml` extension, `.yaml` extension, and `.js` extension.

**Comment by rumpl** [comment]
> Sorry it took me so long, I finally had time today to implement this.

Please don't be sorry, this is awesome! :)

File: README.md [hunk 2]
```diff
@@ -79,6 +79,19 @@ All of the arguments are optional:
 
 `--parsers`, `--detectors` and `--specials`: These arguments are for advanced usage. They provide an easy way to customize the file parser and dependency detection. Check [the pluggable design document](https://github.com/depcheck/depcheck/blob/master/doc/pluggable-design.md) for more information.
 
+## Usage with a configuration file
+
+Depcheck can be used along with rc configuration files. In order to do so, create a .depcheckrc file in your project's package.json folder, and set the CLI keys in YAML format.
+For example, the CLI arguments `--ignores="eslint,babel-*" --skip-missing=true` would turn into:
```
Comment (43081j) [inline README.md]: worth mentioning here what formats are supported rather than just YAML (as i think you said it can do JSON too?)

Comment (Urik) [reply README.md]: Good catch, updated

**Comment by azai91** [comment]
Any idea when this will be merged?

**Comment by rumpl** [comment]
Tomorrow, thanks for pinging me, I forgot about this awesome PR

**Comment by VincentLanglet** [comment]
> Tomorrow, thanks for pinging me, I forgot about this awesome PR

@rumpl Tomorrow was yesterday, did you need another ping ? ;)

**Comment by rumpl** [comment]
As soin that conflicts are treated ans codecov is happy :)

**Comment by rumpl** [comment]
@Urik do you think you will have the time to finish this PR?

**Comment by azai91** [comment]
Any update on the status of this PR?

**Comment by Urik** [comment]
@rumpl didn't realize about the coverage! 
I just moved to Montreal and have been super busy, but next week's weekend I'll find some time to check the tests.

**Comment by rumpl** [comment]
Montreal, awesome! Take your time, whenever is good for you is good for me :) Just wanted to know if you still wanted to work on it still or not.

Keep warm, the winter over there can be quite cold I hear :)

**Comment by Urik** [comment]
@rumpl @43081j 
Finally settled and motivated to do some at home stuff. I've rebased, added tests, and fixed formating. Believe it's good for review now.

**Comment by rumpl** [comment]
Oooh, a christmas gift?! Thank you so much, will review quickly

File: src/utils/configuration-reader.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+function parseCsvArray(value) {
+  return (value || '').split(',');
+}
+
+export function getCliArgs(args, version) {
+  return yargs(args)
+    .usage('Usage: $0 [DIRECTORY]')
+    .boolean(['dev', 'ignore-bin-package', 'skip-missing'])
+    .default({
+      dev: true,
```
Comment (rumpl) [inline src/utils/configuration-reader.js]: `dev` was deprecated and removed

File: README.md [hunk 4]
```diff
@@ -82,6 +82,30 @@ All of the arguments are optional:
...
+***.depcheckrc***
+```
+ignores: ["eslint", "babel-*"]
+skip-missing: true
+```
+
+**Important:** if provided CLI arguments conflict with configuration file ones, the CLI ones will take precedence over the rc file ones.
+
+The rc configuration file can also contain the following extensions: `.json`, `.yaml`, `.yml`, `.js`, `.config.js`.
+If the `.js` or `.config.js` extension are used, then the file should export a Javascript object:
```
Comment (rumpl) [inline README.md]: Is there a way to tell cosmiconfig not to load javascript files? I really want the configuration to be static

File: src/utils/configuration-reader.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      'ignore-bin-package': false,
+      'skip-missing': false,
```
Comment (znarf) [inline src/utils/configuration-reader.js]: If we set a default in the CLI parameters here, we will never be able to set the values from the configuration file.

Fix: the default values of the CLI parameters should be removed, they're not necessary.

**Comment by rumpl** [comment]
@Urik, @znarf made some pretty good points [in his comment on his PR](https://github.com/depcheck/depcheck/pull/491#issuecomment-569513164)

File: src/utils/configuration-reader.js (hunk 5 above)
Comment (znarf) [reply src/utils/configuration-reader.js]: `defaultOptions` has the same purpose later and it's great.

File: test/configuration-reader.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+      let thrownException = null;
+      try {
+        await configurationReaderModule.getConfiguration([], 'test', '1.0');
+      } catch (err) {
+        thrownException = err;
+      }
+      // eslint-disable-next-line no-unused-expressions
+      thrownException.should.not.be.null;
+      thrownException.should.be.instanceof(ConfigurationParsingException);
+      // const configResult = await configurationReaderModule.getConfiguration([], test, '1.0');
```
Comment (znarf) [inline test/configuration-reader.js]: Why is that commented? Work in progress or candidate for deletion?

Comment (Urik) [reply test/configuration-reader.js]: Leftover manual test, thanks for the catch.

File: README.md (hunk 4 above)
Comment (Urik) [reply README.md]: Yeah, will be pushed with my next change.

**Comment by Urik** [comment]
@rumpl @znarf I agree that's a good change, we should move `getParsers` from cli.js into the CLI configuration reader.
I'll have that tomorrow.

**Comment by rumpl** [comment]
Everyone happy new year!

**Comment by Urik** [comment]
Happy new years guys! 
I implemented the changes @znarf suggested for the `parsers` usage in the configuration file, removed the `default` calls in Yargs since we deal with that in `defaultOptions`, and removed support to `.js` config files.

**Comment by rumpl** [comment]
Happy new year to you too @Urik !

**Comment by rumpl** [comment]
🎉🎉🎉🎉🎉 THANK YOU 🎉🎉🎉🎉🎉🎉

Thank you so much @Urik for this PR!

**Comment by krystofwoldrich** [comment]
Not sure if this is the right place to ask, but anyway, is there any support for `extends` functionallity? For example I would have base file .depcheckrc.base.json and then .depcheckrc.json which would extend the base file?

**Pull request merged**